### ISSUE TITLE
test_demo.py test hangs

### ISSUE THREAD

**Issue by justin-stephenson** [pbrezina/sssd-tests-poc#5]
Running the command from the crash course, it hangs here:

~~~
(.venv) [justin@fedora sssd-tests-poc]$ pytest --multihost-config ./mhc.yaml --multihost-log-path=./log -v ./tests/test_demo.py                                                               
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.10.6, pytest-7.1.3, pluggy-1.0.0 -- /home/justin/github/sssd-tests-poc/.venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/justin/github/sssd-tests-poc, configfile: pytest.ini
plugins: multihost-3.4
collected 47 items                                                                                                                                                                           

tests/test_demo.py::test__01 (ldap) PASSED                                                                                                                                             [  2%]
tests/test_demo.py::test__02 (ldap) PASSED                                                                                                                                             [  4%]
tests/test_demo.py::test__03 (ldap) PASSED                                                                                                                                             [  6%]
tests/test_demo.py::test__04 (ldap) PASSED                                                                                                                                             [  8%]
tests/test_demo.py::test__05 (ldap) PASSED                                                                                                                                             [ 10%]
tests/test_demo.py::test__06 (ldap) PASSED                                                                                                                                             [ 12%]
tests/test_demo.py::test__07 (ldap) PASSED                                                                                                                                             [ 14%]
tests/test_demo.py::test__08 (ldap) PASSED                                                                                                                                             [ 17%]
tests/test_demo.py::test__09[su] (ldap) PASSED                                                                                                                                         [ 19%]
tests/test_demo.py::test__09[ssh] (ldap) PASSED                                                                                                                                        [ 21%]
tests/test_demo.py::test__10 (ldap) PASSED                                                                                                                                             [ 23%]
tests/test_demo.py::test__11 (ldap) PASSED                                                                                                                                             [ 25%]
tests/test_demo.py::test__12 (ldap) PASSED                                                                                                                                             [ 27%]
tests/test_demo.py::test__13 (ldap) PASSED                                                                                                                                             [ 29%]
tests/test_demo.py::test__14 (ldap) PASSED                                                                                                                                             [ 31%]
tests/test_demo.py::test__15 (ipa) PASSED                                                                                                                                              [ 34%]
tests/test_demo.py::test__16 (ipa) PASSED                                                                                                                                              [ 36%]
tests/test_demo.py::test__17 (ipa) PASSED                                                                                                                                              [ 38%]
tests/test_demo.py::test__18 (ipa) PASSED                                                                                                                                              [ 40%]
tests/test_demo.py::test__19 (ipa) PASSED                                                                                                                                              [ 42%]
tests/test_demo.py::test__20 (ipa) PASSED                                                                                                                                              [ 44%]
tests/test_demo.py::test__21 (ipa) PASSED                                                                                                                                              [ 46%]
tests/test_demo.py::test__22[su] (ipa) PASSED                                                                                                                                          [ 48%]
tests/test_demo.py::test__22[ssh] (ipa) PASSED                                                                                                                                         [ 51%]
tests/test_demo.py::test__23 (ipa) PASSED                                                                                                                                              [ 53%]
tests/test_demo.py::test__24 (ipa) PASSED                                                                                                                                              [ 55%]
~~~

I had to ctrl-C to exit out, once that is done when I try to re-run the tests they all fail because the directory /etc/systemd/system/sssd.service.d is not cleaned up.

~~~                                                                                                                                                                                                                                                                                
lib/multihost/plugin/fixtures.py:55:                                                                                                                                                                                                                                            
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                                                   
lib/multihost/multihost.py:137: in __enter__                                                                                                                                                                                                                                    
    self._setup()                                                                                                                                                                                                                                                               
lib/multihost/multihost.py:116: in _setup                                                                                                                                                                                                                                       
    role.setup()                                                                                                                                                                                                                                                                
lib/multihost/roles/client.py:29: in setup                                                                                                                                                                                                                                      
    super().setup()                                                                                                                                                                                                                                                             
lib/multihost/roles/base.py:55: in setup                                                                                                                                                                                                                                        
    MultihostUtility.SetupUtilityAttributes(self)                                                                                                                                                                                                                               
lib/multihost/utils/base.py:62: in SetupUtilityAttributes                                                                                                                                                                                                                       
    util.setup()                                                                                                                                                                                                                                                                
lib/multihost/utils/sssd.py:46: in setup                                                                                                                                                                                                                                        
    self.fs.mkdir('/etc/systemd/system/sssd.service.d')                                                                                                                                                                                                                         
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                                                   
                                                                                                                                                                                                                                                                                
self = <lib.multihost.utils.fs.HostFileSystem object at 0x7f3c107160b0>, path = '/etc/systemd/system/sssd.service.d'                                                                                                                                                            
                                                                                                                                                                                                                                                                                
    def mkdir(self, path: str, *, mode: str = None, user: str = None, group: str = None) -> None:                                                                                                                                                                               
        """                                                                                                                                                                                   
        Create directory on remote host.                                                                                                                                                      
                                                                                                                                                                                              
        :param path: Path of the directory.                                                                                                                                                   
        :type path: str                                                                                                                                                                       
        :param mode: Access mode (chmod value), defaults to None                                                                                                                              
        :type mode: str, optional                                                                                                                                                             
        :param user: Owner, defaults to None                                                                                                                                                  
        :type user: str, optional                                                                                                                                                             
        :param group: Group, defaults to None                                                                                                                                                 
        :type group: str, optional                                                                                                                                                            
        :raises OSError: If directory can not be created.                                                                                                                                     
        """                                                                                                                                                                                   
        cmd = f'''                                                                                                                                                                            
        set -x                                                                                                                                                                                
                                                                                                                                                                                              
        mkdir '{path}'                                                                                                                                                                        
        {self.__gen_chattrs(path, mode=mode, user=user, group=group)}                                                                                                                         
        '''                                                                                                                                                                                   
                                                                                                                                                                                              
        result = self.host.exec(cmd, raise_on_error=False)                                                                                                                                    
        if result.rc != 0:                                                                                                                                                                    
>           raise OSError(result.stderr)                   
E           OSError: + mkdir /etc/systemd/system/sssd.service.d                                                        
E           mkdir: cannot create directory ‘/etc/systemd/system/sssd.service.d’: File exists                           

lib/multihost/utils/fs.py:60: OSError                               
~~~

**Comment by pbrezina** [comment]
Can you provide logs?

**Comment by justin-stephenson** [comment]
This morning it hangs at a different test, but the same issue. Log attached. [log.txt](https://github.com/pbrezina/sssd-tests-poc/files/9617722/log.txt)

~~~
[justin@fedora sssd-tests-poc]$ tail log 
pytest_multihost.host.Host.master.cmd11 - DEBUG - Temporary setting umask to 022
pytest_multihost.host.Host.master.cmd11 - DEBUG - Each master will individually need to be re-initialized or
pytest_multihost.host.Host.master.cmd11 - DEBUG - re-created from this one. The replication agreements on
pytest_multihost.host.Host.master.cmd11 - DEBUG - masters running IPA 3.1 or earlier will need to be manually
pytest_multihost.host.Host.master.cmd11 - DEBUG - re-enabled. See the man page for details.
pytest_multihost.host.Host.master.cmd11 - DEBUG - Disabling all replication.
pytest_multihost.host.Host.master.cmd11 - DEBUG - Starting Directory Server
pytest_multihost.host.Host.master.cmd11 - DEBUG - Restoring from userRoot in IPA-TEST
pytest_multihost.host.Host.master.cmd11 - DEBUG - Waiting for LDIF to finish
pytest_multihost.host.Host.master.cmd11 - DEBUG - Restoring from ipaca in IPA-TEST
~~~

**Comment by justin-stephenson** [comment]
Going through the crash course tasks (I am on Fedora 35) and this happens often with IPA tests, including a simple one such as:

~~~
@pytest.mark.topology(KnownTopology.IPA)
def test__21(client: Client, ipa: IPA):
    ipa.user('tuser').add(password='Secret123')

    client.sssd.start()
    assert client.auth.ssh.password('tuser', 'Secret123')
~~~

Tail end of the log is:
~~~
pytest_multihost.host.Host.master.cmd6 - DEBUG - RUN ['ipa-restore', '--unattended', '--password', 'Secret123', '--data', '--online', 'ipa-data-2022-09-21-17-21-49']
pytest_multihost.host.Host.master.cmd6 - DEBUG - Preparing restore from /var/lib/ipa/backup/ipa-data-2022-09-21-17-21-49 on master.ipa.test
pytest_multihost.host.Host.master.cmd6 - DEBUG - Performing DATA restore from DATA backup
pytest_multihost.host.Host.master.cmd6 - DEBUG - Temporary setting umask to 022
pytest_multihost.host.Host.master.cmd6 - DEBUG - Each master will individually need to be re-initialized or
pytest_multihost.host.Host.master.cmd6 - DEBUG - re-created from this one. The replication agreements on
pytest_multihost.host.Host.master.cmd6 - DEBUG - masters running IPA 3.1 or earlier will need to be manually
pytest_multihost.host.Host.master.cmd6 - DEBUG - re-enabled. See the man page for details.
pytest_multihost.host.Host.master.cmd6 - DEBUG - Disabling all replication.
pytest_multihost.host.Host.master.cmd6 - DEBUG - Starting Directory Server
pytest_multihost.host.Host.master.cmd6 - DEBUG - Restoring from userRoot in IPA-TEST
pytest_multihost.host.Host.master.cmd6 - DEBUG - Waiting for LDIF to finish
pytest_multihost.host.Host.master.cmd6 - DEBUG - Restoring from ipaca in IPA-TEST
~~~

**Comment by pbrezina** [comment]
This has never happened to me. This step takes several seconds (~5 on my machine), but if it really hangs then it will be a bug in ipa-restore.

This is log of successful run:
```
pytest_multihost.host.Host.master.cmd8 - DEBUG - RUN ['ipa-restore', '--unattended', '--password', 'Secret123', '--data', '--online', 'ipa-data-2022-09-21-14-40-28']
pytest_multihost.host.Host.master.cmd8 - DEBUG - Preparing restore from /var/lib/ipa/backup/ipa-data-2022-09-21-14-40-28 on master.ipa.test
pytest_multihost.host.Host.master.cmd8 - DEBUG - Performing DATA restore from DATA backup
pytest_multihost.host.Host.master.cmd8 - DEBUG - Temporary setting umask to 022
pytest_multihost.host.Host.master.cmd8 - DEBUG - Each master will individually need to be re-initialized or
pytest_multihost.host.Host.master.cmd8 - DEBUG - re-created from this one. The replication agreements on
pytest_multihost.host.Host.master.cmd8 - DEBUG - masters running IPA 3.1 or earlier will need to be manually
pytest_multihost.host.Host.master.cmd8 - DEBUG - re-enabled. See the man page for details.
pytest_multihost.host.Host.master.cmd8 - DEBUG - Disabling all replication.
pytest_multihost.host.Host.master.cmd8 - DEBUG - Starting Directory Server
pytest_multihost.host.Host.master.cmd8 - DEBUG - Restoring from userRoot in IPA-TEST
pytest_multihost.host.Host.master.cmd8 - DEBUG - Waiting for LDIF to finish
pytest_multihost.host.Host.master.cmd8 - DEBUG - Restoring from ipaca in IPA-TEST
pytest_multihost.host.Host.master.cmd8 - DEBUG - Waiting for LDIF to finish
pytest_multihost.host.Host.master.cmd8 - DEBUG - Restoring umask to 18
pytest_multihost.host.Host.master.cmd8 - DEBUG - The ipa-restore command was successful
pytest_multihost.host.Host.master.cmd8 - DEBUG - Exit code: 0
```

The command still continues after ipaca. So it really hangs inside ipa-restore. Can you please debug it with someone from IPA?

**Comment by justin-stephenson** [comment]
I will file a bug but I don't know if it will get any traction, we will see if any others have this same problem or not as it happens often for me.

~~~
[12:37:16] <tbordaz> justintime: the import thread is stuck because it needs a locker that is acquired by another thread
[12:37:17] <tbordaz>     12d2 dd=15 locks held 1    write locks 0    pid/thread 99/139968453752384 flags 10   priority 100      
[12:37:18] <tbordaz>     12d2 READ          1 HELD    ipaca/entryusn.db         handle        0
[12:37:18] <tbordaz>     12d3 dd=14 locks held 0    write locks 0    pid/thread 99/139968867460672 flags 0    priority 100      
[12:37:18] <tbordaz>     12d3 WRITE         1 WAIT    ipaca/entryusn.db         handle        0
[12:37:55] <tbordaz> justintime: the problem is that the thread holding the locker is no longer running !
[12:39:21] <tbordaz> justintime: This "ghost" thread is likely another task that ran before
[12:40:31] <tbordaz> justintime: something noticeable is that this "ghost" thread held many lockers on files from 'userRoot', and only entryusn.db from ipaca. 
[12:41:16] <tbordaz> justintime: It looks that a thread left the DB quite in a bad shape
[12:41:45] <tbordaz> justintime: I would hope that before leaving it logged some traces in the error logs
[12:54:17] <justintime> tbordaz, thanks for the information, this issue happens in an environment where ipa-backup and ipa-restore are called often times in quick succession - probably this missing thread is from a previous backup/restore operation?
[12:56:23] <tbordaz> justintime: yes, administrative tasks stepping on each other looks the most probable explanation
[12:58:09] <tbordaz> justintime: anyway, a thread leaving without releasing lockers is a bug
~~~

**Comment by pbrezina** [comment]
It has never happened to me, so there must be something in your environment that affects it if it happens often.

So it is deadlock in the directory server?

**Comment by justin-stephenson** [comment]
> It has never happened to me, so there must be something in your environment that affects it if it happens often.
> 
> So it is deadlock in the directory server?

I guess so, I created https://github.com/389ds/389-ds-base/issues/5467
